### PR TITLE
[16.0][MIG] sale_start_end_dates

### PR DESCRIPTION
**Pull request by vancouver29**

### CODE REVIEW

**Comment by rousseldenis** [comment]
@vancouver29 Thanks for this. Could you follow migration guide in order to have two commits (one for pre-commit, one for migration). https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0

File: sale_start_end_dates/models/sale_order.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def _inverse_number_of_days(self):
                     number_of_days=line.number_of_days,
                 )
                 line.number_of_days = 1
-            if line.start_date:
+            if line.start_date and not line.end_date:
```
Comment (rousseldenis) [inline sale_start_end_dates/models/sale_order.py]: Shouldn't the end date converted into computed field readonly False ?

File: sale_start_end_dates/models/sale_order.py [hunk 2]
```diff
@@ -160,12 +160,12 @@ def start_end_dates_product_id_change(self):
         if self.product_id.must_have_dates:
             if self.order_id.default_start_date:
                 self.start_date = self.order_id.default_start_date
-            else:
-                self.start_date = False
+            # else:
```
Comment (rousseldenis) [inline sale_start_end_dates/models/sale_order.py]: Please remove unused code

**Review by rousseldenis — CHANGES_REQUESTED** [review]

**Comment by rousseldenis** [comment]
@vancouver29 Could you also convert your PR title to human readable one ? Like '[16.0][MIG] sale_start_end_dates'. Thanks

**Comment by vancouver29** [comment]
@rousseldenis FYI

**Review by rousseldenis — APPROVED** [review]
Code review

File: sale_start_end_dates/tests/test_sale_start_end_dates.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        self.so.order_line[0].number_of_days = 1
+        self.assertEqual(
+            self.so.order_line[0].start_date, self.so.order_line[0].end_date
+        )
+        self.so.order_line[0].number_of_days = -1
+        self.assertEqual(self.so.order_line[0].number_of_days, 1)
+
+    def test_prepare_invoice_line(self):
+        invoice_line_vals = self.so.order_line._prepare_invoice_line()
+        self.assertEqual(invoice_line_vals["product_id"], self.product_id.id)
```
Comment (chandni299) [inline sale_start_end_dates/tests/test_sale_start_end_dates.py:122]: we can add two more lines here to check `start_date` and `end_date` on invoice line

```
self.assertEqual(invoice_line_vals["start_date"], self.so.order_line[0].start_date)
self.assertEqual(invoice_line_vals["end_date"], self.so.order_line[0].end_date)
```

Comment (vancouver29) [reply sale_start_end_dates/tests/test_sale_start_end_dates.py:122]: @chandni299 i added these lines

**Review by chandni299 — APPROVED** [review]
Looks Good :+1:

**Comment by DorianMAG** [comment]
Hi, looks good to me.
Tested on local environment

**Comment by DorianMAG** [comment]
Maybe, can you rebase your branch with main branch?
Regards

**Comment by alexis-via** [comment]
I improved this PR by converting onchange to compute methods, cf this other PR https://github.com/OCA/sale-workflow/pull/2577
So I propose to close this PR.

**Comment by rousseldenis** [comment]
@vancouver29 Do you mind if we switch to #2577 ?

**Comment by alexis-via** [comment]
The module has been merged in v16, we can now close this PR

**Pull request closed without merging**